modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/MappingQueryContext.java

```diff
@@ -45,4 +53,16 @@ public UUID localNodeId() {
     public AffinityTopologyVersion topologyVersion() {
         return topVer;
     }
+
+    /** Creates a cluster. */
+    RelOptCluster cluster() {
+        if (cluster == null) {
+            cluster = RelOptCluster.create(Commons.emptyCluster().getPlanner(), Commons.emptyCluster().getRexBuilder());
```

Comment:
Seems, VolcanoPlanner can cache schemas (see VolcanoPlanner.registerSchema).
Would you mind to fix the planner as well?

We have similar fix in Ignite-3 (commit d87f1e18) with the method stub and tests looks fine.
But I've tried to apply the same to your PR, but it lead to 2 tests (in PlannerTest) failure with no obvious reason.

Reply:
The planner shouldn't be used by the empty cluster at all, there are assertions to prevent this (I've looked at the fix to ignite-3, and looks like the same check was implemented but in a different way, by prohibiting registerSchema invocation). I can add the same check to the planner as well, but now from my point of view, it looks redundant.
The current fix prohibits not only planning but also mapping (mapping use metadataQuerySupplier, I'm not sure mapping uses planner somehow). 
Also, I think the fix for ignite-3 is not complete, we should attach fragments to the empty cluster again before storing it   to `executionPlan` variable (see https://github.com/apache/ignite/pull/9758/files#diff-07fea6ecf0f238df45e8d6fdffabf83cf9a5fdbd90b31f897868a282c0694f0eR99), otherwise cluster with cached metadata can't be garbage collected (referenced from query cache -> query template -> execution plan -> fragments -> cluster)

Reply:
>  I can add the same check to the planner as well, but now from my point of view, it looks redundant.
I've tried to add the check to your PR code and found 2 tests failed because planner registers schema in empty cluster.
(empty Cluster -> VolcanoPlanner -> registeredSchemas).

> Also, I think the fix for ignite-3 is not complete, we should attach fragments to the empty cluster again before storing it to executionPlan variable

Looks like you are right.

Reply:
> I've tried to add the check to your PR code and found 2 tests failed because planner registers schema in empty cluster. (empty Cluster -> VolcanoPlanner -> registeredSchemas).

Hmm, when I add assert to the `registerSchema` there are a lot of tests failed, because this method is invoked in TableScan rel creation, and this rel is created even when attached to the empty cluster (while cloning rels, while creating it on remote nodes from JSON, etc). I override this method for the empty planner as no-op, please have a look again.